### PR TITLE
first commit

### PR DESCRIPTION
**Pull request by guillaumecrespel**
Not testing

### CODE REVIEW

File: src/roadhog.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+const addQueryParams = (url, queryParams) => {
+  const params = Object.keys(queryParams).map((k => `${k}=${queryParams[k]}`))
+  return `${url}?${join(params, '&')}`
+}
+
+export default action => function* (params) {
+  const pathParams = (params && params.pathParams) || []
+  const queryParams = (params && params.queryParams) || {}
+
+  let url = null
```
Comment (fabienjuif) [inline src/roadhog.js]: Why do you use `null` ?

File: src/roadhog.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  let fallback = undefined
+
+  // Check url redux
+  if (typeof action === 'string') {
+    if (/.*_.*/.test(action)) {
+      // retrieve resource urls.
+      const splitAction = action.split(/_(.+)/)
```
Comment (fabienjuif) [inline src/roadhog.js]: `const [method, resource] = action.split(/_(.+)/)`

File: src/roadhog.js [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+   ,:;;:::;;;;;; ;;;;;;;;;;;;;;;;;; ;;;;;;:::;::
+    ::::::;;;;.   .;;;;;;;;;;;;;;    :;;;::::::,
+     :::::;;`        ,;;;;;;;;.        ,;:::::`
+      .::,                               .:::
+*/
+import { select } from 'redux-saga/effects'
+import { join } from 'lodash'
+import tracer from './tracer'
+
+// Reducer config > api is mandatory.
```
Comment (fabienjuif) [inline src/roadhog.js]: I think we should use jsdoc for this kind of API.

File: src/README.MD [hunk 4]
```diff
@@ -0,0 +1,26 @@
+FETCH (WINSTON) :
```
Comment (fabienjuif) [inline src/README.MD]: Move it to issues

File: package.json [hunk 5]
```diff
@@ -0,0 +1,93 @@
+{
+  "name": "roadhog-saga",
+  "version": "0.0.2",
+  "description": "Library to fetch api, connect to redux, and dispatch saga event",
```
Comment (fabienjuif) [inline package.json]: 💩

File: src/roadhog.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+const apiSelector = ({ config: { api } }) => api
+
+// Add all params to path url.
+const addPathParams = (url, pathParams) => {
+  if (Array.isArray(pathParams) && pathParams.length > 0) return `${url}/${pathParams.join('/')}`
+  return url
+}
+// Add all params to the query on url.
+const addQueryParams = (url, queryParams) => {
+  if (typeof queryParams === 'object' && Object.keys(queryParams).length > 0) {
```
Comment (fabienjuif) [inline src/roadhog.js]: You are calling `Object.keys` twice.
I prefer using a constant

File: src/roadhog.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+*/
+import { select } from 'redux-saga/effects'
+import tracer from './tracer'
+
+// Reducer config > api is mandatory.
+const apiSelector = ({ config: { api } }) => api
+
+// Add all params to path url.
+const addPathParams = (url, pathParams) => {
+  if (Array.isArray(pathParams) && pathParams.length > 0) return `${url}/${pathParams.join('/')}`
```
Comment (fabienjuif) [inline src/roadhog.js]: Why are you testing type ?

File: src/roadhog.js (hunk 6 above)
Comment (fabienjuif) [inline src/roadhog.js]: Why are you testing type ?

Comment (guillaumecrespel) [inline src/roadhog.js]: Yes Sire !

File: src/roadhog.js [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+}
+// Add all params to the query on url.
+const addQueryParams = (url, queryParams) => {
+  if (typeof queryParams === 'object' && Object.keys(queryParams).length > 0) {
+    const params = Object.keys(queryParams).map((k => `${k}=${queryParams[k]}`))
+    return `${url}?${params.join('&')}`
+  }
+  return url
+}
+
```
Comment (fabienjuif) [inline src/roadhog.js]: JSDOC

File: src/roadhog.js [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    }
+  }
+
+  // build url with path params.
+  url = addPathParams(url, pathParams)
+  // build url with query params.
+  url = addQueryParams(url, queryParams)
+
+  // Call tracer : fetch resource and dispatch event error - if necessary -
+  const raw = yield tracer(action, () => fetch(url), !fallback)()
```
Comment (fabienjuif) [inline src/roadhog.js]: You don't try/catch tracer?

File: src/roadhog.js [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+      const resource = api[resource][method]
+      if (typeof resource === 'string') url = resource
+      if (typeof resource === 'object') {
+        url = resource.url
+        fallback = resource.fallback
+      }
+    } else {
+      // throw Exception if action key is malformed.
+      throw new Error(`The action '${action}' is malformed, the template of action is like this 'METHOD_RESOURCES' (ie: GET_USERS)`)
```
Comment (kshish2) [inline src/roadhog.js]: Wrong format for action: '${action}'. Should be "METHOD(GET/PUT)"_"RESOURCE".

File: src/tracer.js [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+ */
+export default (name, callback, sendError) => function* (args) {
+  let raw
+  try {
+    // Start Event
+    yield put(apiEvent(name, 'STARTED'))
+    raw = yield callback(args)
+    // Push saga event.
+    if (!raw.ok && sendError) {
+      throw new Error(`the fetch response is on error : ${raw.status} - ${raw.statusText}`)
```
Comment (kshish2) [inline src/tracer.js]: has an error ?

File: src/tracer.js [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+                                  ``              ``
+*/
+import { put } from 'redux-saga/effects'
+
+// Event catch by saga with template start by 'API_'
+const apiEvent = (name, suffix, payload) => ({ type: `API_${name}_${suffix}`, payload })
+
+/*
+ Tracer :
+ @param {string} name - use for identify saga event.
```
Comment (kshish2) [inline src/tracer.js]: used to identify the saga event

File: src/tracer.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+                  -hNNNmddNd.`+/////:.`  ````  `./////// smo
+                 :ddhs/-./do` /////////:-....-:////////: hs:
+                `/-.`    .o ` -+////////+/////////////+. d-`
+                         ``   `/////////-.``.-///////+: `d
+                               `:///:-.        .-:///-` `+
+                                  ``              ``
+*/
+import { put } from 'redux-saga/effects'
+
+// Event catch by saga with template start by 'API_'
```
Comment (kshish2) [inline src/tracer.js]: Event cought by saga with template starting with 'API_'

File: README.md [hunk 14]
```diff
@@ -1,2 +1,28 @@
 # roadhog
 Fetch your api easily
+
+## Utilisation d�sir�e:
```
Comment (fabienjuif) [inline README.md]: issues

File: src/tracer.js [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+                               `:///:-.        .-:///-` `+
+                                  ``              ``
+*/
+import { put } from 'redux-saga/effects'
+
+// Event caught by saga with template starting with 'API_'
+const apiEvent = (name, suffix, payload) => ({ type: `API_${name}_${suffix}`, payload })
+
+/*
+ Tracer :
```
Comment (fabienjuif) [inline src/tracer.js]: You are missing the * on the column

**Pull request merged**